### PR TITLE
rangefeed: prioritize system spans in rangefeed scheduler

### LINKED ISSUE #110567: rangefeed: add scheduler priority shard for system ranges
To ensure system ranges don't get head-of-line blocked by user rangefeed traffic, we can add a separate priority shard in the scheduler for these, similarly to what we do in the Raft scheduler.

A couple of wrinkles: this will require a `sync.IntMap` lookup for every enqueue, which may have non-negligible cost, and we may not know whether a range is a system range when registering the rangefeed because span configs are themselves propagated via rangefeeds.

Related to #110344.

Jira issue: CRDB-31508

Epic CRDB-26372

### PR DESCRIPTION
**Pull request by erikgrinaker**
**rangefeed: add support for priority processors in scheduler**

This adds a separate shard with a dedicated worker pool that can be used for priority processors, to avoid head-of-line blocking. This is intended for use with system ranges, which are low-volume but need to be low-latency.

**kvserver: prioritize system spans in rangefeed scheduler**

This patch marks system spans as having priority in the rangefeed scheduler. We currently do this based on the descriptor ID, although this no longer strictly holds with dynamic system table IDs.

Resolves #110567.
Release note: None

### CODE REVIEW

**Comment by erikgrinaker** [comment]
Using a priority shard doesn't have measurable impact on rangefeed throughput. Wdyt @aliher1911, worth wrapping this up and getting it in?

```
name                                                         old time/op  new time/op  delta
Rangefeed/procType=scheduler/opType=write/numRegs=1-24       1.33µs ± 3%  1.33µs ± 3%    ~     (p=0.670 n=10+10)
Rangefeed/procType=scheduler/opType=write/numRegs=10-24      8.95µs ± 2%  8.65µs ± 4%  -3.32%  (p=0.004 n=9+10)
Rangefeed/procType=scheduler/opType=write/numRegs=100-24      131µs ± 2%   131µs ± 2%    ~     (p=0.315 n=10+9)
Rangefeed/procType=scheduler/opType=commit/numRegs=1-24      3.60µs ± 2%  3.59µs ± 3%    ~     (p=0.853 n=10+10)
Rangefeed/procType=scheduler/opType=commit/numRegs=10-24     10.7µs ± 4%  10.4µs ± 2%  -2.96%  (p=0.006 n=10+8)
Rangefeed/procType=scheduler/opType=commit/numRegs=100-24     134µs ± 2%   135µs ± 3%    ~     (p=0.971 n=10+10)
Rangefeed/procType=scheduler/opType=closedts/numRegs=1-24    1.03µs ± 6%  1.06µs ± 5%    ~     (p=0.247 n=10+10)
Rangefeed/procType=scheduler/opType=closedts/numRegs=10-24   7.67µs ± 6%  7.63µs ± 2%    ~     (p=0.481 n=10+10)
Rangefeed/procType=scheduler/opType=closedts/numRegs=100-24   121µs ± 1%   121µs ± 1%    ~     (p=0.200 n=8+9)
```

**Comment by erikgrinaker** [comment]
>Do you have an idea how we would determine priority eligibility?

Yeah, I just did this for now. Opened #110883 for a better solution, I think we need to plumb it down from the client.

File: pkg/kv/kvserver/rangefeed/scheduler_test.go [hunk 1]
```diff
@@ -520,3 +532,86 @@ func TestNewSchedulerShards(t *testing.T) {
 		})
 	}
 }
+
+// TestSchedulerPriority tests that the scheduler correctly registers
+// and enqueues events for priority processors.
+func TestSchedulerPriority(t *testing.T) {
```
Comment (aliher1911) [inline pkg/kv/kvserver/rangefeed/scheduler_test.go:538]: We can reuse consumer that is used in other tests. We only need to either add priority to create consumer or have separate method to create with priority:

```
func TestProcessPriority(t *testing.T) {
	defer leaktest.AfterTest(t)()
	ctx := context.Background()
	stopper := stop.NewStopper()
	defer stopper.Stop(ctx)

	s := NewScheduler(SchedulerConfig{Workers: 1, PriorityWorkers: 1, ShardSize: 1})
	require.NoError(t, s.Start(ctx, stopper), "failed to start")
	c1 := createAndRegisterConsumerOrFail(t, s)
	c2 := createAndRegisterPriorityConsumerOrFail(t, s)
	c1.pause()
	s.Enqueue(c1.id, te1)
	c1.waitPaused()
	s.Enqueue(c2.id, te1)
	c2.requireHistory(t, time.Second*30, []processorEventType{te1})
	c1.resume()
	c1.requireHistory(t, time.Second*30, []processorEventType{te1})
	assertStopsWithinTimeout(t, s)
	c1.requireStopped(t, time.Second*30)
	c2.requireStopped(t, time.Second*30)
}
```
Basically copy-pasta of test above that verifies that one blocked consumer doesn't prevent the other to proceed.

**Review by aliher1911 — APPROVED** [review]
LGTM!

**Comment by knz** [comment]
Can we register additional spans as priority system spans after initialization?

**Comment by erikgrinaker** [comment]
>Can we register additional spans as priority system spans after initialization?

The scheduler supports it, but we hardcode the policy based on `MaxReservedDescID`. There's a follow-up issue to generalize this in #110883, probably by letting the client decide the prioritization. Longer-term, we may need smarter prioritization, since we e.g. don't dynamically scale the priority worker pool based on the number of rangefeeds.

**Comment by erikgrinaker** [comment]
We might be able to determine if it was sent from a SQL pod or KV pod.

**Comment by aliher1911** [comment]
If we have this as part of zone config we could restart ranges if we see that they are running with incorrect priority. We can add this functionality to gradual restart part.

File: pkg/kv/kvserver/rangefeed/scheduler.go [hunk 2]
```diff
@@ -160,7 +178,14 @@ func NewScheduler(cfg SchedulerConfig) *Scheduler {
 
 	s := &Scheduler{}
 
-	// Create shards.
+	// Priority shard at index 0.
+	priorityWorkers := 1
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/scheduler.go:182]: Being able to use 1 is useful e.g. in tests. The CRDB default is 2, and is never dynamically adjusted.

**Pull request merged**